### PR TITLE
Watchdog action that will signal a particular thread to abort.

### PR DESCRIPTION
**Pull request by KBaichoo**
Signed-off-by: Kevin Baichoo <kbaichoo@google.com>

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Commit Message: Watchdog action that will signal a particular thread to abort.
Additional Description:
Risk Level: medium
Testing: Unit tests
Docs Changes: Included
Release Notes: Included
Issue: #11388

### CODE REVIEW

File: test/extensions/watchdog/abort_action/abort_action_test.cc
Comment (antoniovicente) [inline test/extensions/watchdog/abort_action/abort_action_test.cc:29]: You may want to restore signal handlers in the test destructor.

File: test/extensions/watchdog/abort_action/abort_action_test.cc
Comment (antoniovicente) [reply test/extensions/watchdog/abort_action/abort_action_test.cc:82]: It seems to me that right now you can't distinguish between death due to the kill signal vs the call to panic.   The kill signal code could be removed and this test would still pass.

File: test/extensions/watchdog/abort_action/abort_action_test.cc
Comment (antoniovicente) [inline test/extensions/watchdog/abort_action/abort_action_test.cc:87]: Is this call to signal(SIGABRT, SIG_IGN); required?

**Review by antoniovicente — APPROVED** [review]

File: test/extensions/watchdog/abort_action/abort_action_test.cc
Comment (KBaichoo) [reply test/extensions/watchdog/abort_action/abort_action_test.cc:87]: Not strictly, but it prevents subsequent printouts of `Eating signal:`

**Comment by KBaichoo** [comment]
PTAL @envoyproxy/senior-maintainers

File: api/envoy/extensions/watchdog/abort_action/v3alpha/abort_action.proto
Comment (ggreenway) [inline api/envoy/extensions/watchdog/abort_action/v3alpha/abort_action.proto:20]: Comment seems wrong/out of date. Note that this comment is the primary documentation for this in the generated docs. It should include some information about what this does, when/why use it, etc.

File: source/extensions/watchdog/abort_action/abort_action.cc
Comment (ggreenway) [inline source/extensions/watchdog/abort_action/abort_action.cc:34]: what is `ltt` stand for? Can you use something more descriptive please?

File: source/extensions/watchdog/abort_action/abort_action.h
Comment (ggreenway) [inline source/extensions/watchdog/abort_action/abort_action.h:15]: Perhaps some/all of this comment block should be included with the proto config, so it gets included in generated docs.

**Review by ggreenway — CHANGES_REQUESTED** [review]
/wait

File: source/extensions/watchdog/abort_action/abort_action.h
Comment (KBaichoo) [reply source/extensions/watchdog/abort_action/abort_action.h:15]: Done. Added additional information from the action's c++ implementation as you suggested.

File: source/extensions/watchdog/abort_action/abort_action.cc
Comment (KBaichoo) [reply source/extensions/watchdog/abort_action/abort_action.cc]: Replaced it with `thread_last_checkin_pairs` hopefully that's a bit more descriptive; it's the last time the thread checked in with the watchdog system. I also changed where `ltt` was used to be `last_checkin` to be a more consistent.

File: api/envoy/extensions/watchdog/abort_action/v3alpha/abort_action.proto
Comment (KBaichoo) [reply api/envoy/extensions/watchdog/abort_action/v3alpha/abort_action.proto]: Done. Added additional information from the action's c++ implementation as you suggested.

File: source/extensions/watchdog/abort_action/abort_action.cc
Comment (ggreenway) [reply source/extensions/watchdog/abort_action/abort_action.cc]: Yes, that's much clearer. Thanks!

File: api/envoy/extensions/watchdog/abort_action/v3alpha/abort_action.proto
Comment (ggreenway) [inline api/envoy/extensions/watchdog/abort_action/v3alpha/abort_action.proto:26]: Should this be the default, on supported platforms? Is there any downside to this? It seems to me that it gives better information (or a chance of it), and the same net behavior (process terminated). @envoyproxy/maintainers

**Review by ggreenway — DISMISSED** [review]
Code LGTM, unless we want to pursue making this the default.

File: api/envoy/extensions/watchdog/abort_action/v3alpha/abort_action.proto
Comment (htuch) [reply api/envoy/extensions/watchdog/abort_action/v3alpha/abort_action.proto:26]: I'm fine with that being default in a future PR.

Comment (htuch) [reply api/envoy/extensions/watchdog/abort_action/v3alpha/abort_action.proto:26]: But also fine here if we want this, seems generally useful.

Comment (KBaichoo) [reply api/envoy/extensions/watchdog/abort_action/v3alpha/abort_action.proto:26]: Sounds good, I can make it be a default action in a future PR.

The main downside I see of making it a default now would be different behaviors across platforms due to a lack of support on Windows. Should we wait for parity before we make it a default?

Comment (ggreenway) [reply api/envoy/extensions/watchdog/abort_action/v3alpha/abort_action.proto:26]: I think it's ok to make it the default, because the only platform difference should be diagnostic output. The process is terminated on all platforms.

I'm also fine with making it the default in a future PR. It's your choice, @KBaichoo .

Comment (KBaichoo) [reply api/envoy/extensions/watchdog/abort_action/v3alpha/abort_action.proto:26]: Sounds good. I'll submit a follow up to this adding this action to the watchdog actions by default.

**Comment by KBaichoo** [comment]
PTAL @envoyproxy/api-shepherds

File: api/envoy/extensions/watchdog/abort_action/v3alpha/abort_action.proto
Comment (mattklein123) [reply api/envoy/extensions/watchdog/abort_action/v3alpha/abort_action.proto:26]: +1 to make this the default.

File: source/extensions/watchdog/abort_action/abort_action.cc
Comment (mattklein123) [inline source/extensions/watchdog/abort_action/abort_action.cc:46]: Is this actually needed? I think Windows has its own extension bzl file so I'm not sure this is even compiled? Could we actually avoid the ifdef in here now or just replace with an proprocessor error if this extension gets compiled on windows?

File: source/extensions/watchdog/abort_action/abort_action.cc
Comment (mattklein123) [inline source/extensions/watchdog/abort_action/abort_action.cc:62]: I haven't kept fully up to date on this, but isn't the default action at the end of the action chain to terminate the process? Do we need this or would the action chain just end up aborting anyway?

**Review by mattklein123 — COMMENTED** [review]
Nice, cool stuff. LGTM with some small comments.

/wait-any

File: source/extensions/watchdog/abort_action/abort_action.cc
Comment (KBaichoo) [reply source/extensions/watchdog/abort_action/abort_action.cc:62]: Yes, you're right. I discussed this with antonio up above. I've copied the reasoning below (copied from https://github.com/envoyproxy/envoy/pull/12860#discussion_r480353221)

There are 3 reasons I think this [having a panic in the action] would be handy (vs deferring to the underlying watchdog default actions[its panic there]):

- Easier to see where the failure occurred, and that signaling failed without having to jump around as much to understand the behavior.
- It's also keeps this self-contained if there were changes on the underlying system behavior in watchdog kill / multikill default
- Allows flexibility for using this in places where there isn't a default PANIC afterwards

Given that we're planning to make this also a default action on platform where it's supported I'd likely make the following change in the next PR:
- If we're doing kill or multikill, then install the abort action as the final watchdog action for either of those events.

I could possibly see when this has full support across platforms that we'd remove the watchdog's default panic, since it'd end up being dead code.

File: source/extensions/watchdog/abort_action/abort_action.cc
Comment (KBaichoo) [reply source/extensions/watchdog/abort_action/abort_action.cc]: Good to know this is a possibility, thanks for pointing it out.

I've added the extension to `WINDOWS_SKIP_TARGETS` which IIUC will let it be skipped by windows so I can remove the `ifdef #Win32` from code.

File: source/extensions/watchdog/abort_action/abort_action.cc
Comment (mattklein123) [reply source/extensions/watchdog/abort_action/abort_action.cc:57]: OK that make sense. Can you summarize this comment in the code? Thank you.

/wait

Comment (KBaichoo) [reply source/extensions/watchdog/abort_action/abort_action.cc]: Done.

File: source/extensions/watchdog/abort_action/abort_action.cc
Comment (KBaichoo) [reply source/extensions/watchdog/abort_action/abort_action.cc]: turns out you also need to add `tags = ["skip_on_windows"]` on tests since excluding the extension in the `.bzl` doesn't effect the tests.

**Comment by mattklein123** [comment]
Can you merge main? It should fix coverage.

**Review by mattklein123 — DISMISSED** [review]

**Comment by mattklein123** [comment]
Sorry coverage failure looks legit. 

Can you check?

/wait

**Comment by KBaichoo** [comment]
the LCOV from the CI is: https://storage.googleapis.com/envoy-pr/12860/coverage/source/extensions/watchdog/abort_action/abort_action.cc.gcov.html 

But if we look at the tests, we do "cover" those lines, but they end up being wrapped in `EXPECT_DEATH` since it's going to kill the process. `EXPECT_DEATH` code paths are run in a sub-process of the main unit testing process. The code coverage tools will never report full coverage for most of the test then.

I will do a per extension percentage override to work around it.

**Review by antoniovicente — DISMISSED** [review]

**Review by mattklein123 — APPROVED** [review]
I think death tests sometimes do count for coverage, but I don't know when/how that happens. It would be nice to fix this at some point but LGTM for now!

**Comment by mattklein123** [comment]
Still broken. :(

/wait

**Comment by KBaichoo** [comment]
Ah, 🤦  the comment lines aren't "covered"... so the percentages have dropped slightly.

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**
